rfc-141-application-healthchecks.md

```diff
@@ -0,0 +1,261 @@
+probably fine.  The problem is elsewhere.
+
+Two of these indicate capacity problems outside of the instance, and
+the third is an alert that a manual maintenance procedure needs to be
+performed some time in the next two months.  They all share the
+property that if one instance reports a failure, all will.  This makes
+them unsuitable for load balancing purposes.
+
+Even if we don't make our healthchecks suitable for load balancing, as
+GOV.UK moves towards continuous deployments, we will end up in
+situations where an automatic deployment is aborted because of
+something unrelated to the change being deployed.  That's not good.
+
+A healthcheck should check that the instance is running and isn't
+completely broken.  That's all.
+
+## Proposal
+
+I propose that we adopt these definitions:
+
+> A *liveness healthcheck* is an HTTP endpoint which MUST return an
+> HTTP status of 200.
+>
+> A *readiness healthcheck* is an HTTP endpoint which MUST return an
+> HTTP status of either 200 (if the instance may receive requests) or
+> 500 (if it should not).  It MAY also return details in the response
+> body as a JSON object.
+
+And furthermore that we commit to deprecating the `/healthcheck`
+endpoint and implementing the new healthchecks at endpoints
+`/healthcheck/live` and `/healthcheck/ready`.
+
+This new approach to readiness healthchecks is more suitable for our
+three purposes than before:
+
+- For load balancing, an unhealthy app will be taken out of the pool.
+- For continuous deployments, we can check the HTTP status rather than
+  read the response body.
+- For alerting, we can continue to read the response body.
+
+We get no immediate benefit from the liveness healthcheck, but will do
+when we have replatformed.
+
+Adopting these definitions gives us some migration work to do.  We
+can't start serving non-"ok" healthchecks with an HTTP status of 500
+right now, as some of our healthchecks are unsuitable for load
+balancing.
+
+We will need to implement the proposal in stages:
+
+1. List all the healthchecks which need changing, because they don't
+   indicate a critical failure of the app (done, see the appendix)
+2. For each such healthcheck:
+   - remove it if it's not adding value, or
+   - add a separate alert if it is
+3. For each application:
+   - serve a liveness healthcheck on `/healthcheck/live`
+   - serve a readiness healthcheck on `/healthcheck/ready`
+4. Update govuk-puppet and govuk-aws to use `/healthcheck/ready`
+   instead of `/healthcheck`
+5. Remove the `/healthcheck` endpoint from every app.
+6. Change `GovukHealthcheck` to serve the appropriate HTTP status
+   code.
+
+As part of (2), we will remove the "warning" state some of our
+healthchecks return.
+
+### Why separate healthchecks?
+
+Separate liveness and readiness healthchecks are a common best
+practice.  They are separate because they serve different purposes:
+
+- Liveness is used by the application runtime (such as [Amazon ECS][],
```

Comment:
Slightly pedantic terminology fix for clarity (sorry to be picky but I think this one's worth it)

```suggestion
- Liveness is used by the container orchestrator (such as [Amazon ECS][],
```

(The specific component of the continainer orchestrator that actually does the liveness probe is typically the _node agent_ (kubelet, borglet, dockerd etc.) but I think that'd be an unnecessary level of detail here.)